airflow/decorators/assets.py

```diff
@@ -0,0 +1,129 @@
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import inspect
+import types
+from typing import TYPE_CHECKING, Any, Iterator, Mapping
+
+import attrs
+
+from airflow.assets import Asset, _validate_identifier
+from airflow.models.dag import DAG, ScheduleArg
+from airflow.operators.python import PythonOperator
+
+if TYPE_CHECKING:
+    from airflow.io.path import ObjectStoragePath
+
+
+@attrs.define(kw_only=True)
+class AssetRef:
+    """Reference to an asset."""
+
+    name: str
+
+
+class _AssetMainOperator(PythonOperator):
+    def __init__(self, *, definition_name: str, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self._definition_name = definition_name
+
+    def _iter_kwargs(self, context: Mapping[str, Any]) -> Iterator[tuple[str, Any]]:
+        for key in inspect.signature(self.python_callable).parameters:
+            if key == "self":
+                value: Any = AssetRef(name=self._definition_name)
+            elif key == "context":
+                value = context
+            else:
+                # TODO: This does not check if the upstream asset actually
+                # exists. Should we do a second pass in the DAG processor to
+                # raise parse-time errors if a non-existent asset is referenced?
+                # How? Should we also fail the task at runtime? Or should the
+                # dangling reference simply do nothing?
+                value = AssetRef(name=key)
+            yield key, value
+
+    def determine_kwargs(self, context: Mapping[str, Any]) -> Mapping[str, Any]:
+        return dict(self._iter_kwargs(context))
+
+
+@attrs.define(kw_only=True)
+class AssetDefinition(Asset):
+    """
+    Asset representation from decorating a function with ``@asset``.
+
+    :meta private:
+    """
+
+    function: types.FunctionType
+    schedule: ScheduleArg
+
+    def __attrs_post_init__(self) -> None:
+        parameters = inspect.signature(self.function).parameters
+        with DAG(dag_id=self.name, schedule=self.schedule, auto_register=True) as dag:
+            _AssetMainOperator(
+                task_id="__main__",
+                inlets=[
+                    Asset(name=inlet_aset_name)
+                    for inlet_aset_name in parameters
+                    if inlet_aset_name not in ("self", "context")
+                ],
```

Comment:
I don’t think this is good enough since `Asset(name="a")` and `Asset(name="a", uri="somewhere")` are considered different assets. Here we only have the name, so maybe one way to do this would be to add a stub (AssetRef?) here, and do a second pass in the DAG processor to get the actual assets? Or maybe we should just keep the AssetRef here, and add code to task execution at runtime to resolve it into an actual asset, similar to how we handle AssetAlias.

Reply:
What are the `actual assets` you mentioned? What should we reference assets if we're using the name `a`?

1. `Asset(name="a")` only
2. All the assets with `name=a` (e.g, `Asset(uri="a")`, `Asset(uri="s3://...", name="a")`, `Asset(uri="gcs://...", name="a")`)

Reply:
Just traced the code a bit and figure out we've not yet handled it in inlet side https://github.com/apache/airflow/blob/06088a3abcbb46533e74de360746db766d50cf66/airflow/utils/context.py#L242

Reply:
```python
@asset(uri="snowflake://.../bus_trips")
def aggregated_bus_trips(raw_bus_trips, context):
    # Write aggregated bus trips to asset...
    print(context["inlet_events"][raw_bus_trips])
```

@uranusjr 
Would also like to confirm by defining something above, we want to get the asset_event of the active asset with the name `raw_bus_trips`. Or is there anything I missed?

Reply:
namely we'll probably need to change the SQL here https://github.com/apache/airflow/blob/06088a3abcbb46533e74de360746db766d50cf66/airflow/utils/context.py#L291

Reply:
also you mentioned the DAG processing thing, is that still needed if I go with the above approach (I guess not?)

Reply:
Yeah this should work for the user. Although I feel we should resolve `raw_bus_trips` inside the function into an Asset object automatically when we execute it, instead of having it as a ref and do special deferecencing inside the accessor. This is because the user may want to use the value for other purposes, and it would likely be more difficult to make things make sense if `raw_bus_trips` is of a different class than Asset, which is the primary interface to the user.

Reply:
If that's the case, we probably should just fetch the active asset here https://github.com/apache/airflow/pull/41325/files#diff-1331de0aa19a45db3dbaaa3aab5cd28a00e9090fc33dfa20d4a7fe75887cb1efR93 ?

If we resolve somewhere around execution https://github.com/apache/airflow/pull/41325/files#diff-1331de0aa19a45db3dbaaa3aab5cd28a00e9090fc33dfa20d4a7fe75887cb1efR66, the `inlets` would still be `AssetRef`. Then we'll need to do it here and also do the deferecencing thing.

Reply:
The values in `inlets` being AssetRef is probably OK; introspecting a task is not a common thing to do (we need to support things like OpenLineage, but that’s not too difficult). We can probably resolve as late as `determine_kwargs`.

Reply:
I'm now resolving it in `determine_kwargs` and `InletEventsAccessors.__getitem__`. The `InletEventsAccessors.__getitem__` can probably be cached but it's not done at this moment.

Reply:
I'm thinking of checking the outlet and inlet thing again in another PR. This might be handled there.